pkg/sst/sst.go

```diff
@@ -234,3 +238,119 @@ func getBits(val, i, j uint32) uint32 {
 func isBitSet(val, n uint32) bool {
 	return val&(1<<n) != 0
 }
+
+func setBit(val, n uint32) uint32 {
+	return val | (1 << n)
+}
+
+func clearBit(val, n uint32) uint32 {
+	return val &^ (1 << n)
+}
+
+func setBFStatus(info *SstPackageInfo, status bool) error {
+	rsp, err := sendMboxCmd(info.pkg.cpus[0], CONFIG_TDP, CONFIG_TDP_GET_TDP_CONTROL, uint32(info.PPCurrentLevel))
+	if err != nil {
+		return fmt.Errorf("failed to read SST status: %v", err)
+	}
+
+	req := clearBit(rsp, 17)
+	if status {
+		req = setBit(rsp, 17)
+	}
+
+	if _, err = sendMboxCmd(info.pkg.cpus[0], CONFIG_TDP, CONFIG_TDP_SET_TDP_CONTROL, req); err != nil {
+		return fmt.Errorf("failed to enable SST %s: %v", "BF", err)
+	}
+
+	info.BFEnabled = status
+
+	return nil
+}
+
+func adjustScalingMinFromBaseFreq(cpu utils.ID) error {
+	minScalingFreq, err := getCPUScalingMinFreq(cpu)
+	if err != nil {
+		sstlog.Debugf("sst-bf cpu %d set scaling freq error: %v", cpu, err)
+		return err
+	}
+
+	baseFreq, err := getCPUBaseFreq(cpu)
+	if err != nil {
+		sstlog.Debugf("sst-bf cpu %d set base freq error: %v", cpu, err)
+		return err
+	}
+
+	// Boost the scaling_min_freq if the base frequency is higher
+	if minScalingFreq < baseFreq {
+		if err := setCPUScalingMinFreq(cpu, baseFreq); err != nil {
+			sstlog.Debugf("sst-bf cpu %d set min scaling freq error: %v", cpu, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func setScalingMin2CPUInfoMax(info *SstPackageInfo) error {
+	for _, cpu := range info.BFCores.Members() {
+		err := setCPUScalingMin2CPUInfoMaxFreq(cpu)
+		if err != nil {
+			sstlog.Debugf("sst-bf cpu %d set min freq error: %v", cpu, err)
+			return err
+		}
+
+		err = adjustScalingMinFromBaseFreq(cpu)
```

Comment:
This is done same way in intel-speed-select tool, see
https://github.com/torvalds/linux/blob/master/tools/power/x86/intel-speed-select/isst-config.c#L1496
for a reference. If the condition is always false, then of course we should remove the check.

Reply:
No point necessarily aimlessly just copying the reference implementation.

Now with better understanding of how this all works we should actually think what would be "the right thing" or the most sensible thing to do with the cpufreq attributes. I'm not quite sure that setting `scaling_min` to `cpuinfo_max` is actually the good action in the general case. Especially in the API

Reply:
I suggest to drop `adjustScalingMinFromBaseFreq`. Unless you can explain how it would work. AFAIU, essentially what we have now is equivalent to this

```
min = 1
base = 5
max = 10

scalingMin = max

if (scalingMin < base) {
   scalingMin = base
}
```